go/arrow/avro/reader_types.go

```diff
@@ -0,0 +1,892 @@
+			case nil:
+				bt.AppendNull()
+			case []interface{}:
+				if len(dt) == 0 {
+					bt.AppendEmptyValue()
+				} else {
+					bt.Append(true)
+				}
+			default:
+				bt.Append(true)
+			}
+			return nil
+		}
+	case *array.ListBuilder:
+		lf := parent.newChild(field.Name, b, field.Metadata)
+		vb := bt.ValueBuilder()
+		lf.isList = true
+		mapFieldBuilders(vb, field.Type.(*arrow.ListType).ElemField(), lf)
+		lf.appendFunc = func(data interface{}) error {
+			switch dt := data.(type) {
+			case nil:
+				bt.AppendNull()
+			case []interface{}:
+				if len(dt) == 0 {
+					bt.AppendEmptyValue()
+				} else {
+					bt.Append(true)
+				}
+			default:
+				bt.Append(true)
+			}
+			return nil
+		}
+	case *array.MapBuilder:
+		// has metadata for objects in values
+		mf := parent.newChild(field.Name, b, field.Metadata)
+		mf.isMap = true
+		kb := bt.KeyBuilder()
+		ib := bt.ItemBuilder()
+		mapFieldBuilders(kb, field.Type.(*arrow.MapType).KeyField(), mf)
+		mapFieldBuilders(ib, field.Type.(*arrow.MapType).ItemField(), mf)
+		mf.appendFunc = func(data interface{}) error {
+			switch data.(type) {
+			case nil:
+				bt.AppendNull()
+			default:
+				bt.Append(true)
+			}
+			return nil
+		}
+	case *array.MonthDayNanoIntervalBuilder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendDurationData(bt, data)
+			return nil
+		}
+	case *array.StringBuilder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendStringData(bt, data)
+			return nil
+		}
+	case *array.StructBuilder:
+		// has metadata for Avro Union named types
+		sf := parent.newChild(field.Name, b, field.Metadata)
+		sf.typeName, _ = field.Metadata.GetValue("typeName")
+		sf.isStruct = true
+		// create children
+		for i, f := range field.Type.(*arrow.StructType).Fields() {
+			mapFieldBuilders(bt.FieldBuilder(i), f, sf)
+		}
+		sf.appendFunc = func(data interface{}) error {
+			switch data.(type) {
+			case nil:
+				bt.AppendNull()
+			default:
+				appendStructData(bt, data)
+			}
+			return nil
+		}
+	case *array.Time32Builder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendTime32Data(bt, data)
+			return nil
+		}
+	case *array.Time64Builder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendTime64Data(bt, data)
+			return nil
+		}
+	case *array.TimestampBuilder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendTimestampData(bt, data)
+			return nil
+		}
+	}
+}
+
+func appendBinaryData(b *array.BinaryBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case map[string]any:
+		switch ct := dt["bytes"].(type) {
+		case nil:
+			b.AppendNull()
+		default:
+			b.Append(ct.([]byte))
+		}
+	default:
+		b.Append([]byte(fmt.Sprint(data)))
```

Comment:
might be *slightly* more performant to do `b.Append(fmt.Append([]byte{}, data))` which would avoid the intermediate string allocation

Reply:
suggestion implemented